# Project: cleanning - [candy data]
### this notebook is for educational purpose . in this notebook we will do some cleanning

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">wrangling</a></li>
  
</ul>

<a id='intro'></a>
## Introduction

### Dataset Description 

> i will go through candy data 2017 , this data has :
- Internal_ID
- Q1-Going Out?: Are you actually going trick or treating yourself? "has values `Yes or No`"
- Q2-Gender: It has four different options, `Femal ,Male ,Other ,I'd rather not say`
- Q3-Age: Numerical field
- Q4-Country: Text Field, but users have written their own version of the names. Example, for America, there are entries such as USA, us, US, America so we should consider that while cleanning
- Q5-State/Province: Text Field, but users have written their own version of the names. Same as the country data.
- Q6-Joy Or Despair: All kinds of chocolate bars are the questions with three distinct options to choose from (Joy, Meh, Despair).
- Q7-Joy Other: Text Field to write items not included above that give you JOY. Lots of missing values.
- Q8-Despair Other: Text Field ti write items not included above that give you DESPAIR. Lots of missing values.
- Q9-Other Comments: Text Field. Lots of missing values.
- Q10-Dress: Binary field. Missing values present
- Unnamed: 113
- Q11-Day: Binary Answer Field. Missing values present


# 1- import packages and load data

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from fuzzywuzzy import process
from fuzzywuzzy import fuzz

pd.options.mode.chained_assignment = None
df = pd.read_excel("candyhierarchy2017.xlsx", encoding='latin-1')
pd.set_option('display.max_columns', None)

<a id='wrangling'></a>
## Data Wrangling



### General Properties

# 2- show some general properties for the data 

In [2]:
df.shape

(2460, 120)

In [3]:
df.iloc[[*range(2), *range(-2, 0)]]

Internal ID Q1: GOING OUT? Q2: GENDER Q3: AGE Q4: COUNTRY  \
0        90258773            NaN        NaN     NaN         NaN   
1        90272821             No       Male      44        USA    
2458     90314658             No       Male      58         Usa   
2459     90314802             No     Female      66         usa   

     Q5: STATE, PROVINCE, COUNTY, ETC Q6 | 100 Grand Bar  \
0                                 NaN                NaN   
1                                  NM                MEH   
2458                   North Carolina                NaN   
2459                     Pennsylvania            DESPAIR   

     Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                                   NaN                                       
1                                               DESPAIR                                       
2458                                                NaN                                       
2459                                            DESPAIR                                       

     Q6 | Any full-sized candy bar Q6 | Black Jacks Q6 | Bonkers (the candy)  \
0                              NaN              NaN                      NaN   
1                              JOY              MEH                  DESPAIR   
2458                           NaN              NaN                      NaN   
2459                           JOY          DESPAIR                  DESPAIR   

     Q6 | Bonkers (the board game) Q6 | Bottle Caps Q6 | Box'o'Raisins  \
0                              NaN              NaN                NaN   
1                          DESPAIR          DESPAIR            DESPAIR   
2458                           NaN              NaN                NaN   
2459                       DESPAIR          DESPAIR                MEH   

     Q6 | Broken glow stick Q6 | Butterfinger Q6 | Cadbury Creme Eggs  \
0                       NaN               NaN                     NaN   
1                   DESPAIR           DESPAIR                     MEH   
2458                    NaN               NaN                     NaN   
2459                DESPAIR               JOY                     MEH   

     Q6 | Candy Corn  \
0                NaN   
1                MEH   
2458             NaN   
2459             MEH   

     Q6 | Candy that is clearly just the stuff given out for free at restaurants  \
0                                                   NaN                            
1                                               DESPAIR                            
2458                                                NaN                            
2459                                            DESPAIR                            

     Q6 | Caramellos Q6 | Cash, or other forms of legal tender  \
0                NaN                                       NaN   
1                MEH                                       JOY   
2458             NaN                                       NaN   
2459             MEH                                       JOY   

     Q6 | Chardonnay Q6 | Chick-o-Sticks (we don’t know what that is)  \
0                NaN                                              NaN   
1                MEH                                          DESPAIR   
2458             NaN                                              NaN   
2459             JOY                                          DESPAIR   

     Q6 | Chiclets Q6 | Coffee Crisp  \
0              NaN               NaN   
1          DESPAIR           DESPAIR   
2458           NaN               NaN   
2459           MEH           DESPAIR   

     Q6 | Creepy Religious comics/Chick Tracts Q6 | Dental paraphenalia  \
0                                          NaN                      NaN   
1                                      DESPAIR                  DESPAIR   
2458                                       NaN                      NaN   
2459           

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2460 entries, 0 to 2459
Columns: 120 entries, Internal ID to Click Coordinates (x, y)
dtypes: float64(4), int64(1), object(115)
memory usage: 2.3+ MB


In [5]:
df.dtypes.iloc[[*range(7), *range(-12, 0)]]

Internal ID                           int64
Q1: GOING OUT?                       object
Q2: GENDER                           object
Q3: AGE                              object
Q4: COUNTRY                          object
Q5: STATE, PROVINCE, COUNTY, ETC     object
Q6 | 100 Grand Bar                   object
Q6 | York Peppermint Patties         object
Q7: JOY OTHER                        object
Q8: DESPAIR OTHER                    object
Q9: OTHER COMMENTS                   object
Q10: DRESS                           object
Unnamed: 113                         object
Q11: DAY                             object
Q12: MEDIA [Daily Dish]             float64
Q12: MEDIA [Science]                float64
Q12: MEDIA [ESPN]                   float64
Q12: MEDIA [Yahoo]                  float64
Click Coordinates (x, y)             object
dtype: object

In [6]:
df.nunique().iloc[[*range(7), *range(-12, 0)]]

Internal ID                         2460
Q1: GOING OUT?                         2
Q2: GENDER                             4
Q3: AGE                              106
Q4: COUNTRY                          128
Q5: STATE, PROVINCE, COUNTY, ETC     511
Q6 | 100 Grand Bar                     3
Q6 | York Peppermint Patties           3
Q7: JOY OTHER                        867
Q8: DESPAIR OTHER                    685
Q9: OTHER COMMENTS                   385
Q10: DRESS                             2
Unnamed: 113                           1
Q11: DAY                               2
Q12: MEDIA [Daily Dish]                1
Q12: MEDIA [Science]                   1
Q12: MEDIA [ESPN]                      1
Q12: MEDIA [Yahoo]                     1
Click Coordinates (x, y)             952
dtype: int64

In [7]:
df.isnull().sum().iloc[[*range(7), *range(-12, 0)]]

Internal ID                            0
Q1: GOING OUT?                       110
Q2: GENDER                            41
Q3: AGE                               84
Q4: COUNTRY                           64
Q5: STATE, PROVINCE, COUNTY, ETC     100
Q6 | 100 Grand Bar                   747
Q6 | York Peppermint Patties         705
Q7: JOY OTHER                       1549
Q8: DESPAIR OTHER                   1743
Q9: OTHER COMMENTS                  2074
Q10: DRESS                           744
Unnamed: 113                        2451
Q11: DAY                             725
Q12: MEDIA [Daily Dish]             2375
Q12: MEDIA [Science]                1098
Q12: MEDIA [ESPN]                   2361
Q12: MEDIA [Yahoo]                  2393
Click Coordinates (x, y)             855
dtype: int64

In [8]:
df.duplicated().sum()

0

In [9]:
df[df.isnull().all(axis=1)].index

Int64Index([], dtype='int64')

# 3- write summary about what you get form the above functions
> #### tips : this summary can be about : if there is any missing values , if there a wrong data type , if there a wrong range of values , ....

# Note :
>Above data has 120 columns where six have str entries (Age, Country, State ,Other...) whereas other 114 has Categorical entries.

>Data has zero duplicate and zero null row.

>Data has 2460 rows and 120 columns.

>Q6 columns are the main columns where this survey is created

>Q7, Q8, Q9 and "Unnamed: 113" columns have much missing values.

# cleaning

### 4- drop all the useless columns and  rows with too much NaN , 
> tip : useless columns can be like `Internal ID`and can be also columns with too much NaN 
- rename columns for easy access

### Droping useless columns

In [10]:
df.drop(["Unnamed: 113", "Internal ID", "Click Coordinates (x, y)", 'Q7: JOY OTHER', 
        'Q8: DESPAIR OTHER', 'Q9: OTHER COMMENTS'] , axis="columns", inplace=True)

### renaming columns

In [11]:
df = df.rename(columns = {'Q1: GOING OUT?' : 'Going Out?',
                          'Q2: GENDER' : 'Gender',
                          'Q3: AGE': 'Age',
                          'Q4: COUNTRY' : 'Country',
                          'Q5: STATE, PROVINCE, COUNTY, ETC' : 'Area',
                          'Q10: DRESS' : 'Dress',
                          'Q11: DAY': 'Day',
                          'Q12: MEDIA [Daily Dish]' : 'DailyDish',
                          'Q12: MEDIA [Science]': 'Science',
                          'Q12: MEDIA [ESPN]' : 'ESPN',
                          'Q12: MEDIA [Yahoo]': 'Yahoo'})

In [12]:
df.columns = [x for x in df.columns.str.replace("Q6 \| ",'')]

In [13]:
df.head(1)

Going Out? Gender  Age Country Area 100 Grand Bar  \
0        NaN    NaN  NaN     NaN  NaN           NaN   

  Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                                NaN                                  

  Any full-sized candy bar Black Jacks Bonkers (the candy)  \
0                      NaN         NaN                 NaN   

  Bonkers (the board game) Bottle Caps Box'o'Raisins Broken glow stick  \
0                      NaN         NaN           NaN               NaN   

  Butterfinger Cadbury Creme Eggs Candy Corn  \
0          NaN                NaN        NaN   

  Candy that is clearly just the stuff given out for free at restaurants  \
0                                                NaN                       

  Caramellos Cash, or other forms of legal tender Chardonnay  \
0        NaN                                  NaN        NaN   

  Chick-o-Sticks (we don’t know what that is) Chiclets Coffee Crisp  \
0                                         NaN      NaN          NaN   

  Creepy Religious comics/Chick Tracts Dental paraphenalia Dots Dove Bars  \
0                                  NaN                 NaN  NaN       NaN   

  Fuzzy Peaches Generic Brand Acetaminophen Glow sticks Goo Goo Clusters  \
0           NaN                         NaN         NaN              NaN   

  Good N' Plenty Gum from baseball cards Gummy Bears straight up Hard Candy  \
0            NaN                     NaN                     NaN        NaN   

  Healthy Fruit Heath Bar Hershey's Dark Chocolate Hershey’s Milk Chocolate  \
0           NaN       NaN                      NaN                      NaN   

  Hershey's Kisses Hugs (actual physical hugs) Jolly Rancher (bad flavor)  \
0              NaN                         NaN                        NaN   

  Jolly Ranchers (good flavor) JoyJoy (Mit Iodine!) Junior Mints Senior Mints  \
0                          NaN                  NaN          NaN          NaN   

  Kale smoothie Kinder Happy Hippo Kit Kat LaffyTaffy LemonHeads  \
0           NaN                NaN     NaN        NaN        NaN   

  Licorice (not black) Licorice (yes black) Lindt Truffle Lollipops Mars  \
0                  NaN                  NaN           NaN       NaN  NaN   

  Maynards Mike and Ike Milk Duds Milky Way Regular M&Ms Peanut M&M’s  \
0      NaN          NaN       NaN       NaN          NaN          NaN   

  Blue M&M's Red M&M's Green Party M&M's Independent M&M's  \
0        NaN       NaN               NaN               NaN   

  Abstained from M&M'ing. Minibags of chips Mint Kisses Mint Juleps  \
0                     NaN               NaN         NaN         NaN   

  Mr. Goodbar Necco Wafers Nerds Nestle Crunch Now'n'Laters Peeps Pencils  \
0         NaN          NaN   NaN           NaN          NaN   NaN     NaN   

  Pixy Stix Real Housewives of Orange County Season 9 Blue-Ray  \
0       NaN                                                NaN   

  Reese’s Peanut Butter Cups Reese's Pieces Reggie Jackson Bar Rolos  \
0                        NaN            NaN                NaN   NaN   

  Sandwich-sized bags filled with BooBerry Crunch Skittles  \
0                                             NaN      NaN   

  Smarties (American) Smarties (Commonwealth) Snickers  \
0                 NaN                     NaN      NaN   

  Sourpatch Kids (i.e. abominations of nature) Spotted Dick Starburst  \
0                                          NaN          NaN       NaN   

  Sweet Tarts Swedish Fish Sweetums (a friend to diabetes) Take 5 Tic Tacs  \
0         NaN          NaN                             NaN    NaN      NaN   

  Those odd marshmallow circus peanut things Three Musketeers  \
0                                        NaN              NaN   

  Tolberone something or other Trail Mix Twix  \
0                          NaN       NaN  NaN   

  Vials of pure high fructose corn syrup, for main-lining into

### Droping useless rows

In [14]:
df.iloc[:,5:-6].columns

Index(['100 Grand Bar',
       'Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)',
       'Any full-sized candy bar', 'Black Jacks', 'Bonkers (the candy)',
       'Bonkers (the board game)', 'Bottle Caps', 'Box'o'Raisins',
       'Broken glow stick', 'Butterfinger',
       ...
       'Three Musketeers', 'Tolberone something or other', 'Trail Mix', 'Twix',
       'Vials of pure high fructose corn syrup, for main-lining into your vein',
       'Vicodin', 'Whatchamacallit Bars', 'White Bread',
       'Whole Wheat anything', 'York Peppermint Patties'],
      dtype='object', length=103)

In [15]:
df[df.iloc[:,5:-6].isnull().all(axis=1)].index

Int64Index([   0,    2,    6,   10,   18,   19,   24,   25,   30,   35,
            ...
            2437, 2440, 2445, 2446, 2447, 2448, 2449, 2450, 2451, 2458],
           dtype='int64', length=657)

In [16]:
df = df.drop(df[df.iloc[:,5:-6].isnull().all(axis=1)].index)

### Resetting the index of the dataframe

In [17]:
df = df.reset_index()
df.drop(['index'] , axis="columns", inplace=True)

In [18]:
df.head(3)

Going Out? Gender Age Country      Area 100 Grand Bar  \
0         No   Male  44    USA         NM           MEH   
1         No   Male  40      us        or           MEH   
2         No   Male  23     usa  exton pa           JOY   

  Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                            DESPAIR                                  
1                                            DESPAIR                                  
2                                            DESPAIR                                  

  Any full-sized candy bar Black Jacks Bonkers (the candy)  \
0                      JOY         MEH             DESPAIR   
1                      JOY         MEH                 MEH   
2                      JOY     DESPAIR                 MEH   

  Bonkers (the board game) Bottle Caps Box'o'Raisins Broken glow stick  \
0                  DESPAIR     DESPAIR       DESPAIR           DESPAIR   
1                  DESPAIR         MEH       DESPAIR           DESPAIR   
2                  DESPAIR         MEH       DESPAIR           DESPAIR   

  Butterfinger Cadbury Creme Eggs Candy Corn  \
0      DESPAIR                MEH        MEH   
1          MEH                MEH    DESPAIR   
2          MEH                MEH    DESPAIR   

  Candy that is clearly just the stuff given out for free at restaurants  \
0                                            DESPAIR                       
1                                            DESPAIR                       
2                                            DESPAIR                       

  Caramellos Cash, or other forms of legal tender Chardonnay  \
0        MEH                                  JOY        MEH   
1        MEH                                  JOY        MEH   
2        JOY                                  MEH        JOY   

  Chick-o-Sticks (we don’t know what that is) Chiclets Coffee Crisp  \
0                                     DESPAIR  DESPAIR      DESPAIR   
1                                         JOY      MEH      DESPAIR   
2                                     DESPAIR  DESPAIR          MEH   

  Creepy Religious comics/Chick Tracts Dental paraphenalia Dots Dove Bars  \
0                              DESPAIR             DESPAIR  MEH       JOY   
1                              DESPAIR             DESPAIR  JOY       JOY   
2                              DESPAIR             DESPAIR  MEH       JOY   

  Fuzzy Peaches Generic Brand Acetaminophen Glow sticks Goo Goo Clusters  \
0       DESPAIR                     DESPAIR     DESPAIR          DESPAIR   
1           MEH                         MEH         JOY              JOY   
2       DESPAIR                     DESPAIR     DESPAIR              MEH   

  Good N' Plenty Gum from baseball cards Gummy Bears straight up Hard Candy  \
0            MEH                 DESPAIR                     MEH        MEH   
1            MEH                 DESPAIR                     JOY        MEH   
2            MEH                 DESPAIR                     JOY        MEH   

  Healthy Fruit Heath Bar Hershey's Dark Chocolate Hershey’s Milk Chocolate  \
0       DESPAIR       MEH                      JOY                      JOY   
1           MEH   DESPAIR                      MEH                      MEH   
2       DESPAIR   DESPAIR                      JOY                      MEH   

  Hershey's Kisses Hugs (actual physical hugs) Jolly Rancher (bad flavor)  \
0              MEH                     DESPAIR                    DESPAIR   
1              MEH                         MEH                        MEH   
2              MEH                     DESPAIR                        MEH   

  Jolly Ranchers (good flavor) JoyJoy (Mit Iodine!) Junior Mints Senior Mints  \
0                          MEH              DESPAIR      DESPAIR      DESPAIR   
1                          MEH              DESPAIR      DESPAIR      DESPAIR   
2                          JOY

### 5 - clean `going out` column 
> tip : you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [19]:
df['Going Out?'].unique()

array(['No', 'Yes', nan], dtype=object)

In [20]:
df['Going Out?'].isnull().sum()

64

In [21]:
df['Going Out?']=df['Going Out?'].fillna('Maybe')
df['Going Out?'].value_counts()

No       1500
Yes       239
Maybe      64
Name: Going Out?, dtype: int64

### 6 - clean ` gender` column 
> tip : you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [22]:
df['Gender'].unique()

array(['Male', 'Female', "I'd rather not say", 'Other', nan], dtype=object)

In [23]:
df['Gender'].isnull().sum()

11

In [24]:
df['Gender']=df['Gender'].fillna("I'd rather not say")
df['Gender'].value_counts()

Male                  1087
Female                 618
I'd rather not say      72
Other                   26
Name: Gender, dtype: int64

### 7 - clean ` age` column 
> tip : note the type of age column , and NaN values
 - try to use `to_numeric` 
 - tip : you should choose which best value to put in place of these NaNs ( mean, median, ....etc)

In [25]:
df['Age'].unique()

array([44, 40, 23, nan, 33, 53, 56, 64, 43, 37, 59, 48, 45, 49, 34, 35,
       38, 58, 50, 47, 52, 63, 65, 41, 31, 61, 46, 54, 62, 29, 42, 39, 36,
       32, 28, 69, 67, 30, 25, 22, 51, 70, 24, 27, 18, 19, 'Old enough',
       57, 66, 12, 'Many', 60, 55, '?', 21, 11, 'no', 68, 20, 6, 10, 71,
       90, 13, 26, '45-55', 312, 72, 99, 'hahahahaha', 9, 88, 39.4, 74,
       102, 'older than dirt', 15, 8, 75, 17, 'Enough', 'See question 2',
       '24-50', 7, 14, 'Over 50', 76, 100, 1000, 16, '46 Halloweens.',
       'ancient', 77, 'OLD', 'old', 'MY NAME JEFF', 4, 73,
       '59 on the day after Halloween', 'old enough',
       'I can remember when Java was a cool new language'], dtype=object)

In [26]:
df['Age']=pd.to_numeric(df['Age'], errors='coerce')

In [27]:
df = df[df['Age'].between(5, 80, inclusive="both")]

In [28]:
df['Age']=df['Age'].fillna(df['Age'].mean())
df['Age']=df['Age'].astype(int)

In [29]:
df['Age'].unique()

array([44, 40, 23, 33, 53, 56, 64, 43, 37, 59, 48, 45, 49, 34, 35, 38, 58,
       50, 47, 52, 63, 65, 41, 31, 61, 46, 54, 62, 29, 42, 39, 36, 32, 28,
       69, 67, 30, 25, 22, 51, 70, 24, 27, 18, 19, 57, 66, 12, 60, 55, 21,
       11, 68, 20,  6, 10, 71, 13, 26, 72,  9, 74, 15,  8, 75, 17,  7, 14,
       76, 16, 77, 73], dtype=int64)

### 8 - clean ` country` column
> tip : note that country column has country names with different writing ways 
  - use fuzzywuzzy to deal with this problem
  - you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [30]:
df['Country'].unique()

array(['USA ', 'us', 'usa', 'canada', 'Canada', 'Us', 'US', 'Murica',
       'USA', 'uk', 'United States', 'united states', 'Usa',
       'United staes', 'United Kingdom', 'United States of America',
       'UAE', 'United States ', 'England', 'canada ', 'United states',
       'u.s.a.', 'USAUSAUSA', 'america', 'france',
       'United States of America ', 'U.S.A.', 'Mexico', 'Canada ',
       'united states of america', 'US of A', 'UK', 'Unites States',
       'The United States', 'North Carolina ', 'Netherlands', 'Europe',
       'U S', 'u.s.', 'Costa Rica', 'The United States of America',
       'cascadia', 'Greece', 'USA? Hard to tell anymore..', "'merica",
       'Pittsburgh', 'United State', 'France', 'U.S.', 'Can', 'Canae',
       'New York', 'Trumpistan', 'Ireland', 'United Sates', 'Korea',
       'Australia', 'California', 'Japan', 'USa', 'South africa',
       'I pretend to be from Canada, but I am really from the United States.',
       'Usa ', 'Iceland', 'Canada`', 'Scotland

In [31]:
df['Country'] = df['Country'].str.strip().str.lower()

In [32]:
countries = pd.Series(['usa', 'united states of america', 'united kingdom england', 'hong kong', 'china', 'sweden', 'spain',
                       'ireland', 'taiwan', 'singapore', 'netherlands', 
                       'indonesia', 'south korea', 'switzerland', 'denmark', 'scotland', 'canada', 'iceland', 'south africa', 
                       'japan', 'greece', 'costa rica', 'germany', 'finland', 'france', 'mexico'])

for country in countries:
    matches = process.extract(country, df['Country'], limit = df.shape[0])
    for potential_match in matches:
        if potential_match[1] >= 80:
            df.loc[df['Country'] == potential_match[0], 'Country'] = country
            
df['Country'].unique()

array(['usa', 'canada', 'murica', 'uk', 'united states of america', 'uae',
       'united kingdom england', 'u.s.a.', 'france', 'mexico',
       'north carolina', 'netherlands', 'europe', 'costa rica',
       'cascadia', 'greece', 'usa? hard to tell anymore..', 'pittsburgh',
       'canae', 'new york', 'trumpistan', 'iceland', 'south korea',
       'australia', 'california', 'japan', 'south africa', 'scotland',
       'denmark', 'switzerland', 'new jersey', 'germany', nan, 'endland',
       'murrika', 'singapore', 'taiwan', 'china', 'spain', 'narnia',
       'u s a', "i don't know anymore"], dtype=object)

In [33]:
df['Country'].value_counts()

usa                            954
united states of america       549
canada                         156
uk                              12
u.s.a.                           8
germany                          7
iceland                          4
netherlands                      4
japan                            3
scotland                         3
france                           3
mexico                           2
south korea                      2
switzerland                      2
denmark                          2
australia                        2
endland                          1
singapore                        1
u s a                            1
china                            1
murica                           1
europe                           1
north carolina                   1
narnia                           1
south africa                     1
uae                              1
murrika                          1
spain                            1
usa? hard to tell an

In [34]:
mapping = {'murica':'other', 'united states of america':'usa', 'u.s.a.':'usa',
'north carolina':'usa', 'usa? hard to tell anymore..':'usa', 'pittsburgh':'usa', 'new york':'usa', 'trumpistan':'usa',
          'california':'usa', 'ud':'usa', 'new jersey':'usa', 'endland':'uk', 'soviet canuckistan':'canada',
           'u s a':'usa', 'united kingdom england':'uk', 'u.k.':'uk', 'cascadia':'usa'}
df['Country'] = df['Country'].replace(mapping)

In [35]:
df.Country[~df['Country'].isin(['usa', 'uk','canada'])]='other'

In [36]:
df['Country'] = df['Country'].str.upper()
df['Country'].value_counts()

USA       1520
CANADA     156
OTHER       49
UK          14
Name: Country, dtype: int64

### 9 - clean ` area` column
> tip : area column has the same problem as country columns , but looking at all it's unique values may give us another way to deal with it rather than try to fix it

In [37]:
df['Area'].nunique()

422

In [38]:
df.drop('Area' , axis="columns", inplace=True)
df.head(1)

Going Out? Gender  Age Country 100 Grand Bar  \
0         No   Male   44     USA           MEH   

  Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                            DESPAIR                                  

  Any full-sized candy bar Black Jacks Bonkers (the candy)  \
0                      JOY         MEH             DESPAIR   

  Bonkers (the board game) Bottle Caps Box'o'Raisins Broken glow stick  \
0                  DESPAIR     DESPAIR       DESPAIR           DESPAIR   

  Butterfinger Cadbury Creme Eggs Candy Corn  \
0      DESPAIR                MEH        MEH   

  Candy that is clearly just the stuff given out for free at restaurants  \
0                                            DESPAIR                       

  Caramellos Cash, or other forms of legal tender Chardonnay  \
0        MEH                                  JOY        MEH   

  Chick-o-Sticks (we don’t know what that is) Chiclets Coffee Crisp  \
0                                     DESPAIR  DESPAIR      DESPAIR   

  Creepy Religious comics/Chick Tracts Dental paraphenalia Dots Dove Bars  \
0                              DESPAIR             DESPAIR  MEH       JOY   

  Fuzzy Peaches Generic Brand Acetaminophen Glow sticks Goo Goo Clusters  \
0       DESPAIR                     DESPAIR     DESPAIR          DESPAIR   

  Good N' Plenty Gum from baseball cards Gummy Bears straight up Hard Candy  \
0            MEH                 DESPAIR                     MEH        MEH   

  Healthy Fruit Heath Bar Hershey's Dark Chocolate Hershey’s Milk Chocolate  \
0       DESPAIR       MEH                      JOY                      JOY   

  Hershey's Kisses Hugs (actual physical hugs) Jolly Rancher (bad flavor)  \
0              MEH                     DESPAIR                    DESPAIR   

  Jolly Ranchers (good flavor) JoyJoy (Mit Iodine!) Junior Mints Senior Mints  \
0                          MEH              DESPAIR      DESPAIR      DESPAIR   

  Kale smoothie Kinder Happy Hippo Kit Kat LaffyTaffy LemonHeads  \
0       DESPAIR            DESPAIR     JOY    DESPAIR        MEH   

  Licorice (not black) Licorice (yes black) Lindt Truffle Lollipops     Mars  \
0                  MEH                  JOY           MEH   DESPAIR  DESPAIR   

  Maynards Mike and Ike Milk Duds Milky Way Regular M&Ms Peanut M&M’s  \
0  DESPAIR          MEH       MEH       JOY          JOY          MEH   

  Blue M&M's Red M&M's Green Party M&M's Independent M&M's  \
0        JOY       JOY               JOY               JOY   

  Abstained from M&M'ing. Minibags of chips Mint Kisses Mint Juleps  \
0                 DESPAIR           DESPAIR         MEH     DESPAIR   

  Mr. Goodbar Necco Wafers    Nerds Nestle Crunch Now'n'Laters    Peeps  \
0     DESPAIR      DESPAIR  DESPAIR           JOY      DESPAIR  DESPAIR   

   Pencils Pixy Stix Real Housewives of Orange County Season 9 Blue-Ray  \
0  DESPAIR   DESPAIR                                            DESPAIR   

  Reese’s Peanut Butter Cups Reese's Pieces Reggie Jackson Bar Rolos  \
0                        JOY            JOY            DESPAIR   JOY   

  Sandwich-sized bags filled with BooBerry Crunch Skittles  \
0                                         DESPAIR  DESPAIR   

  Smarties (American) Smarties (Commonwealth) Snickers  \
0             DESPAIR                 DESPAIR      MEH   

  Sourpatch Kids (i.e. abominations of nature) Spotted Dick Starburst  \
0                                      DESPAIR      DESPAIR       MEH   

  Sweet Tarts Swedish Fish Sweetums (a friend to diabetes)   Take 5 Tic Tacs  \
0     DESPAIR          MEH                         DESPAIR  DESPAIR  DESPAIR   

  Those odd marshmallow circus peanut things Three Musketeers  \
0                                    DESPAIR              JOY   

  Tolberone something or other Trail Mix Twix  \
0                          JOY   DESPAIR  JOY   

  Vials of pure high fructose corn syrup, for 

### 10 - clean ` Q6` columns
>- tip : Q6 has all kinds of chocolate bars and questions with three distinct options to choose from (Joy, Meh, Despair).
- so all the column that has Q6 is a candy name , so you should go through all these column and you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [39]:
df.iloc[:,4:-6].columns

Index(['100 Grand Bar',
       'Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)',
       'Any full-sized candy bar', 'Black Jacks', 'Bonkers (the candy)',
       'Bonkers (the board game)', 'Bottle Caps', 'Box'o'Raisins',
       'Broken glow stick', 'Butterfinger',
       ...
       'Three Musketeers', 'Tolberone something or other', 'Trail Mix', 'Twix',
       'Vials of pure high fructose corn syrup, for main-lining into your vein',
       'Vicodin', 'Whatchamacallit Bars', 'White Bread',
       'Whole Wheat anything', 'York Peppermint Patties'],
      dtype='object', length=103)

In [40]:
df.iloc[:,4:-6]=df.iloc[:,4:-6].fillna('MEH')

In [41]:
df.head(3)

Going Out? Gender  Age Country 100 Grand Bar  \
0         No   Male   44     USA           MEH   
1         No   Male   40     USA           MEH   
2         No   Male   23     USA           JOY   

  Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                            DESPAIR                                  
1                                            DESPAIR                                  
2                                            DESPAIR                                  

  Any full-sized candy bar Black Jacks Bonkers (the candy)  \
0                      JOY         MEH             DESPAIR   
1                      JOY         MEH                 MEH   
2                      JOY     DESPAIR                 MEH   

  Bonkers (the board game) Bottle Caps Box'o'Raisins Broken glow stick  \
0                  DESPAIR     DESPAIR       DESPAIR           DESPAIR   
1                  DESPAIR         MEH       DESPAIR           DESPAIR   
2                  DESPAIR         MEH       DESPAIR           DESPAIR   

  Butterfinger Cadbury Creme Eggs Candy Corn  \
0      DESPAIR                MEH        MEH   
1          MEH                MEH    DESPAIR   
2          MEH                MEH    DESPAIR   

  Candy that is clearly just the stuff given out for free at restaurants  \
0                                            DESPAIR                       
1                                            DESPAIR                       
2                                            DESPAIR                       

  Caramellos Cash, or other forms of legal tender Chardonnay  \
0        MEH                                  JOY        MEH   
1        MEH                                  JOY        MEH   
2        JOY                                  MEH        JOY   

  Chick-o-Sticks (we don’t know what that is) Chiclets Coffee Crisp  \
0                                     DESPAIR  DESPAIR      DESPAIR   
1                                         JOY      MEH      DESPAIR   
2                                     DESPAIR  DESPAIR          MEH   

  Creepy Religious comics/Chick Tracts Dental paraphenalia Dots Dove Bars  \
0                              DESPAIR             DESPAIR  MEH       JOY   
1                              DESPAIR             DESPAIR  JOY       JOY   
2                              DESPAIR             DESPAIR  MEH       JOY   

  Fuzzy Peaches Generic Brand Acetaminophen Glow sticks Goo Goo Clusters  \
0       DESPAIR                     DESPAIR     DESPAIR          DESPAIR   
1           MEH                         MEH         JOY              JOY   
2       DESPAIR                     DESPAIR     DESPAIR              MEH   

  Good N' Plenty Gum from baseball cards Gummy Bears straight up Hard Candy  \
0            MEH                 DESPAIR                     MEH        MEH   
1            MEH                 DESPAIR                     JOY        MEH   
2            MEH                 DESPAIR                     JOY        MEH   

  Healthy Fruit Heath Bar Hershey's Dark Chocolate Hershey’s Milk Chocolate  \
0       DESPAIR       MEH                      JOY                      JOY   
1           MEH   DESPAIR                      MEH                      MEH   
2       DESPAIR   DESPAIR                      JOY                      MEH   

  Hershey's Kisses Hugs (actual physical hugs) Jolly Rancher (bad flavor)  \
0              MEH                     DESPAIR                    DESPAIR   
1              MEH                         MEH                        MEH   
2              MEH                     DESPAIR                        MEH   

  Jolly Ranchers (good flavor) JoyJoy (Mit Iodine!) Junior Mints Senior Mints  \
0                          MEH              DESPAIR      DESPAIR      DESPAIR   
1                          MEH              DESPAIR      DESPAIR      DESPAIR   
2                          JOY                  MEH          JOY  

### 11 - clean ` dress` column
>tip : you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [42]:
df['Dress'].unique()

array(['White and gold', 'Blue and black', nan], dtype=object)

In [43]:
df['Dress']=df['Dress'].fillna('Other')

### 12 - clean ` day` column

In [44]:
df['Day'].unique()

array(['Sunday', 'Friday', nan], dtype=object)

In [45]:
df['Day']=df['Day'].fillna('Other')

### 12 - clean ` Media` column

In [46]:
media = df.iloc[:,-4:]
for i in media.columns:
    media.loc[media[i] == 1.0, i] = i
df['Media'] = media[media.columns[:]].apply(lambda x: ''.join(x.dropna().astype(str)),axis=1)
df['Media']=df['Media'].astype('category')
df.drop(['DailyDish', 'Science', 'ESPN', 'Yahoo'] , axis="columns", inplace=True)
df.loc[df['Media'] == '', 'Media'] = "Science"
df.head(1)

Going Out? Gender  Age Country 100 Grand Bar  \
0         No   Male   44     USA           MEH   

  Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                            DESPAIR                                  

  Any full-sized candy bar Black Jacks Bonkers (the candy)  \
0                      JOY         MEH             DESPAIR   

  Bonkers (the board game) Bottle Caps Box'o'Raisins Broken glow stick  \
0                  DESPAIR     DESPAIR       DESPAIR           DESPAIR   

  Butterfinger Cadbury Creme Eggs Candy Corn  \
0      DESPAIR                MEH        MEH   

  Candy that is clearly just the stuff given out for free at restaurants  \
0                                            DESPAIR                       

  Caramellos Cash, or other forms of legal tender Chardonnay  \
0        MEH                                  JOY        MEH   

  Chick-o-Sticks (we don’t know what that is) Chiclets Coffee Crisp  \
0                                     DESPAIR  DESPAIR      DESPAIR   

  Creepy Religious comics/Chick Tracts Dental paraphenalia Dots Dove Bars  \
0                              DESPAIR             DESPAIR  MEH       JOY   

  Fuzzy Peaches Generic Brand Acetaminophen Glow sticks Goo Goo Clusters  \
0       DESPAIR                     DESPAIR     DESPAIR          DESPAIR   

  Good N' Plenty Gum from baseball cards Gummy Bears straight up Hard Candy  \
0            MEH                 DESPAIR                     MEH        MEH   

  Healthy Fruit Heath Bar Hershey's Dark Chocolate Hershey’s Milk Chocolate  \
0       DESPAIR       MEH                      JOY                      JOY   

  Hershey's Kisses Hugs (actual physical hugs) Jolly Rancher (bad flavor)  \
0              MEH                     DESPAIR                    DESPAIR   

  Jolly Ranchers (good flavor) JoyJoy (Mit Iodine!) Junior Mints Senior Mints  \
0                          MEH              DESPAIR      DESPAIR      DESPAIR   

  Kale smoothie Kinder Happy Hippo Kit Kat LaffyTaffy LemonHeads  \
0       DESPAIR            DESPAIR     JOY    DESPAIR        MEH   

  Licorice (not black) Licorice (yes black) Lindt Truffle Lollipops     Mars  \
0                  MEH                  JOY           MEH   DESPAIR  DESPAIR   

  Maynards Mike and Ike Milk Duds Milky Way Regular M&Ms Peanut M&M’s  \
0  DESPAIR          MEH       MEH       JOY          JOY          MEH   

  Blue M&M's Red M&M's Green Party M&M's Independent M&M's  \
0        JOY       JOY               JOY               JOY   

  Abstained from M&M'ing. Minibags of chips Mint Kisses Mint Juleps  \
0                 DESPAIR           DESPAIR         MEH     DESPAIR   

  Mr. Goodbar Necco Wafers    Nerds Nestle Crunch Now'n'Laters    Peeps  \
0     DESPAIR      DESPAIR  DESPAIR           JOY      DESPAIR  DESPAIR   

   Pencils Pixy Stix Real Housewives of Orange County Season 9 Blue-Ray  \
0  DESPAIR   DESPAIR                                            DESPAIR   

  Reese’s Peanut Butter Cups Reese's Pieces Reggie Jackson Bar Rolos  \
0                        JOY            JOY            DESPAIR   JOY   

  Sandwich-sized bags filled with BooBerry Crunch Skittles  \
0                                         DESPAIR  DESPAIR   

  Smarties (American) Smarties (Commonwealth) Snickers  \
0             DESPAIR                 DESPAIR      MEH   

  Sourpatch Kids (i.e. abominations of nature) Spotted Dick Starburst  \
0                                      DESPAIR      DESPAIR       MEH   

  Sweet Tarts Swedish Fish Sweetums (a friend to diabetes)   Take 5 Tic Tacs  \
0     DESPAIR          MEH                         DESPAIR  DESPAIR  DESPAIR   

  Those odd marshmallow circus peanut things Three Musketeers  \
0                                    DESPAIR              JOY   

  Tolberone something or other Trail Mix Twix  \
0                          JOY   DESPAIR  JOY   

  Vials of pure high fructose corn syrup, for 

In [47]:
df['Media'].unique()

[Science, DailyDish, Yahoo, ESPN]
Categories (4, object): [Science, DailyDish, Yahoo, ESPN]

### 13 - check
> after finish cleaning you should make sure your data is clean and that :
- there is no wrong data type
- there is no NaN values 
- all the column are cleaing and ready for the next step `EDA`

In [48]:
df.dtypes.iloc[[*range(7), *range(-7, 0)]]

Going Out?                                                                             object
Gender                                                                                 object
Age                                                                                     int32
Country                                                                                object
100 Grand Bar                                                                          object
Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)      object
Any full-sized candy bar                                                               object
Whatchamacallit Bars                                                                   object
White Bread                                                                            object
Whole Wheat anything                                                                   object
York Peppermint Patties                                     

In [49]:
for i in df.columns[:]:
    if i == 'Age':
        continue
    df = df.astype({i: 'category'})

In [50]:
df.dtypes.iloc[[*range(7), *range(-7, 0)]]

Going Out?                                                                           category
Gender                                                                               category
Age                                                                                     int32
Country                                                                              category
100 Grand Bar                                                                        category
Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)    category
Any full-sized candy bar                                                             category
Whatchamacallit Bars                                                                 category
White Bread                                                                          category
Whole Wheat anything                                                                 category
York Peppermint Patties                                     

In [51]:
df.isnull().values.any()

False

In [52]:
df.shape

(1739, 110)

### A summary about all the cleaning steps 
>We took each column separately and fill the `NaN` values with the appropriate match.

>Some columns weren't important so droping them with much easier and usefull than trying to clean them.

>I mixed the last foun columns into one column called Media.

>Overall there is now 110 columns including 103 which are the main candy columns.

>I lost almost 700 rows in cleaning.

# Abdelrahman Mostafa